src/wiring_diagrams/ScheduleUndirected.jl

```diff
@@ -0,0 +1,219 @@
+      get!(jmap, j) do; length(jmap) + 1 end
+    end
+    njunc = length(jmap)
+    f(values, map(js -> FinFunction([jmap[j] for j in js], njunc), juncs),
+      FinFunction([jmap[j] for j in outer_junc], njunc))
+  end
+
+  # Mutually recursively evaluate children of composite `c`.
+  function eval_children(c::Int)
+    bs, cs = box_children(d, c), children(d, c)
+    values = [ generators[bs]; map(eval_composite, cs) ]
+    juncs = [ [junction(d, ports(d, b)) for b in bs];
+              [composite_junction(d, composite_ports(d, c′)) for c′ in cs] ]
+    (values, juncs)
+  end
+
+  # Mutually recursively evaluate composite `c`.
+  function eval_composite(c::Int)
+    values, juncs = eval_children(c)
+    outer_junc = composite_junction(d, composite_ports(d, c))
+    do_eval(values, juncs, outer_junc)
+  end
+
+  # Evaluate diagram starting at the root, assumed to be unique. The root
+  # composite must be handled specially due to outer ports.
+  root = only(filter(c -> parent(d, c) == c, composites(d)))
+  values, juncs = eval_children(root)
+  do_eval(values, juncs, junction(d, outer=true))
+end
+
+""" Convert a scheduled UWD to a nested UWD.
+"""
+function to_nested_diagram(s::AbstractScheduledUWD)
+  d = NestedUWD()
+  copy_parts!(d, s)
+
+  n = nboxes(s)
+  sets = IntDisjointSets(n + ncomposites(s))
+
+  function add_composite_ports!(c::Int)
+    # Recursively add ports to all child composites.
+    foreach(add_composite_ports!, children(d, c))
+
+    # Get all junctions incident to any child box or child composite.
+    js = unique!(sort!(flat([
+      [ junction(d, ports(d, b)) for b in box_children(d, c) ];
+      [ composite_junction(d, composite_ports(d, c′)) for c′ in children(d, c) ]
+    ])))
+
+    # Filter for "outgoing" junctions, namely those incident to any outer port
+    # or to any port of a box that is not a descendant of this composite.
+    c_rep = n+c
+    for b in box_children(d, c); union!(sets, c_rep, b) end
+    for c′ in children(d, c); union!(sets, c_rep, n+c′) end
+    js = filter!(js) do j
+      !(all(in_same_set(sets, c_rep, box(d, port))
+            for port in ports_with_junction(d, j)) &&
+        isempty(ports_with_junction(d, j, outer=true)))
+    end
+
+    # Add port for each outgoing junction.
+    add_parts!(d, :CompositePort, length(js), composite=c, composite_junction=js)
+  end
+
+  # Add ports to each composite, starting at roots.
+  roots = filter(c -> parent(d, c) == c, composites(d))
+  foreach(add_composite_ports!, roots)
+  return d
+end
+
+# Scheduling
+############
+
+""" Schedule UWD as a sequential chain of binary composites.
+
+This is the simplest possible scheduling algorithm, the equivalent of `foldl`
+for undirected wiring diagrams. Unless otherwise specified, the boxes are folded
+according to the order of their IDs, which is arbitrary.
+"""
+sequential_schedule(d::AbstractUWD) = sequential_schedule(d, boxes(d))
```

Comment:
This seems like an ideal place to use multiple dispatch to select the scheduling algorithm. 

```julia
abstract type AbstractSchedule end 
struct SequentialSchedule <: AbstractSchedule end
struct BisectionSchedule <: AbstractSchedule end 

function schedule(d::AbstractUWD, alg::AbstractSchedule=SequentialSchedule())
    schedule(d, alg)
end

function schedule(d::AbstractUWD, alg::SequentialSchedule)
    ...
end

function schedule(d::AbstractUWD, alg::BisectionSchedule)
    # schedule by recursive bisection compose(1:floor(Int, n/2), floor(Int, n/2)+1:end)
end
```

Reply:
What would be gained by using multiple dispatch here instead of just having different functions? Seems to me that this just moves the names of algorithms from functions to types. But maybe there is some advantage to that?

Reply:
Oh, I see your comment above about other packages defining their own scheduling algorithms. But for that they can just write functions that take a UWD and output a `ScheduledUWD`, right? Then again, I don't see why DiffEqs does this either, so maybe I am just missing something.

Reply:
Another advantage is that you can specify a default scheduling algorithm. That's good enough for me, I suppose. I'll make this change.